apps/test-app/frontend/src/components/tree-widget/TreeWidget.tsx

```diff
@@ -172,21 +177,32 @@ export function StatelessTreeWidget(props: Omit<Props, "rulesetId">) {
       const parent: HierarchyNode | undefined = node ? (node as any).__internal : undefined;
       try {
         if (modelsTreeHierarchyProvider) {
-          return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent })).map(parseTreeNodeItem);
+          return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent, hierarchyLevelSizeLimit })).map(parseTreeNodeItem);
         }
         return [];
       } catch (e) {
         // eslint-disable-next-line no-console
         console.error(e);
+        if (ErrorTypeChecker.isRowsLimitExceededError(e)) {
+          return [
+            createInfoNode(
+              node,
+              `${IModelApp.localization.getLocalizedString("Sample:controls.result-limit-exceeded")} ${hierarchyLevelSizeLimit!}`,
+              InfoTreeNodeItemType.ResultSetTooLarge,
+            ),
+          ];
+        }
         return [];
       }
     };
-  }, [modelsTreeHierarchyProvider]);
+  }, [modelsTreeHierarchyProvider, hierarchyLevelSizeLimit]);
   const modelSource = useTreeModelSource(dataProvider);
   const nodeLoader = useTreeNodeLoader(dataProvider, modelSource);
   const eventHandler = useMemo(() => new TreeEventHandler({ nodeLoader, modelSource }), [nodeLoader, modelSource]);
   const treeModel = useTreeModel(modelSource);
-
+  const nodeRenderer = useCallback((nodeProps: TreeNodeRendererProps) => {
+    return <StatelessTreeNodeRenderer {...nodeProps} onLimitReset={() => setHierarchyLevelSizeLimit(undefined)} />;
```

Comment:
I don't think this works as discussed. Here's how it should work:
1. By default we send all hierarchy level requests without limit override (and it defaults to 1000).
2. In case of "result set too large" error, we should render a custom node, which offers **to remove the limit for that particular hierarchy level**. Note that removing the limit means there is not limit rather than the default limit.

Reply:
Added option to set `hierarchyLevelSizeLimit` to "unbounded" which allows this functionality.